archiver/service/archiver.go

```diff
@@ -0,0 +1,248 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	client "github.com/attestantio/go-eth2-client"
+	"github.com/attestantio/go-eth2-client/api"
+	v1 "github.com/attestantio/go-eth2-client/api/v1"
+	"github.com/base-org/blob-archiver/archiver/flags"
+	"github.com/base-org/blob-archiver/archiver/metrics"
+	"github.com/base-org/blob-archiver/common/storage"
+	"github.com/ethereum-optimism/optimism/op-service/httputil"
+	opmetrics "github.com/ethereum-optimism/optimism/op-service/metrics"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const backfillErrorRetryInterval = 5 * time.Second
+
+var ErrAlreadyStopped = errors.New("already stopped")
+
+type BeaconClient interface {
+	client.BlobSidecarsProvider
+	client.BeaconBlockHeadersProvider
+}
+
+func NewService(l log.Logger, cfg flags.ArchiverConfig, api *API, dataStoreClient storage.DataStore, client BeaconClient, m metrics.Metricer) (*ArchiverService, error) {
+	return &ArchiverService{
+		log:             l,
+		cfg:             cfg,
+		dataStoreClient: dataStoreClient,
+		metrics:         m,
+		stopCh:          make(chan struct{}),
+		beaconClient:    client,
+		api:             api,
+	}, nil
+}
+
+type ArchiverService struct {
+	stopped         atomic.Bool
+	stopCh          chan struct{}
+	log             log.Logger
+	dataStoreClient storage.DataStore
+	beaconClient    BeaconClient
+	metricsServer   *httputil.HTTPServer
+	cfg             flags.ArchiverConfig
+	metrics         metrics.Metricer
+	api             *API
+}
+
+// Start starts the archiver service. It begins polling the beacon node for the latest blocks and persisting blobs for
+// them. Concurrently it'll also begin a backfill process (see backfillBlobs) to store all blobs from the current head
+// to the previously stored blocks. This ensures that during restarts or outages of an archiver, any gaps will be
+// filled in.
```

Comment:
What happens if the backfill period exceeds the L1 retention window?

Reply:
It'll treat the beacon node of the source of truth and keep walking back, writing empty blobs until it hits the last known block.

I wonder if we should extend this to not fetch blobs older than the blob expiry window (though we would have to make it configurable as some people may adjust the pruning time on their CL nodes).

Reply:
Right we don't know how the beacon node is configured. I think this is fine for now, just make sure the behavior is documented.  We can only archive what's there :)